spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -113,21 +117,42 @@ public DatastoreEntityConverter getDatastoreEntityConverter() {
 
 	@Override
 	public <T> T findById(Object id, Class<T> entityClass) {
-		Iterator<T> results = findAllById(Collections.singleton(id), entityClass)
-				.iterator();
+		return findById(id, entityClass, new ReadContext());
+	}
+
+	private  <T> T findById(Object id, Class<T> entityClass, ReadContext context) {
+		Iterator<T> results = findAllById(Collections.singleton(id), entityClass, context).iterator();
 		return results.hasNext() ? results.next() : null;
 	}
 
 	@Override
 	public <T> T save(T instance, Key... ancestors) {
-		getDatastoreReadWriter().put(convertToEntityForSave(instance, ancestors));
-		return instance;
+		return save(instance, new HashSet<>(), ancestors);
+	}
+
+	private <T> T save(T instance, Set<Key> persisted, Key... ancestors) {
+		Iterable<T> iterable = saveAll(Collections.singletonList(instance), persisted, ancestors);
+		return iterable.iterator().next();
 	}
 
 	@Override
 	public <T> Iterable<T> saveAll(Iterable<T> entities, Key... ancestors) {
-		getDatastoreReadWriter().put(StreamSupport.stream(entities.spliterator(), false)
-				.map((entity) -> convertToEntityForSave(entity, ancestors)).toArray(Entity[]::new));
+		return saveAll(entities, new HashSet<>(), ancestors);
+	}
+
+	private <T> Iterable<T> saveAll(Iterable<T> entities, Set<Key> persisted, Key... ancestors) {
+		List<Entity> entitiesForSave = new ArrayList<>();
+		for (T entity : entities) {
+			Key key = getKey(entity, true, ancestors);
+			if (persisted.contains(key)) {
+				continue;
+			}
+			persisted.add(key);
+			entitiesForSave.add(convertToEntityForSave(entity, persisted, ancestors));
+		}
+		if (!entitiesForSave.isEmpty()) {
+			getDatastoreReadWriter().put(entitiesForSave.toArray(new Entity[0]));
+		}
 		return entities;
```

Comment:
This might be going back a bit, but what's the point of returning the same variable that is sent in as the argument??
Wouldn't returning the actual saved `Entity[]` be more useful?

Reply:
The Spring Data Repository's save method has a return type that is `T` (the original pojo type). The template save method's return type has always been the same to support the repo's save method.

Reply:
True. Do you prefer making that method return nothing? Returning `Entity[]` isn't more useful.

Reply:
Returning Entity I think would be useful if the public method also returned Entity[] or List<Entity>

Reply:
There might be some use-cases, but let's tackle this in a different PR. Can you please create an issue for this?